src/pkgs.jl

```diff
@@ -313,41 +314,46 @@ function add_require(sourcefile, modcaller, idmod, modname, expr)
         while !haskey(pkgdatas, id)
             sleep(0.1)
         end
-        # Get/create the FileInfo specifically for tracking @require blocks
-        pkgdata = pkgdatas[id]
-        filekey = relpath(sourcefile, pkgdata) * "__@require__"
-        fileidx = fileindex(pkgdata, filekey)
-        if fileidx === nothing
-            files = srcfiles(pkgdata)
-            fileidx = length(files) + 1
-            push!(files, filekey)
-            push!(pkgdata.fileinfos, FileInfo(modcaller))
-        end
-        fi = pkgdata.fileinfos[fileidx]
-        # Tag the expr to ensure it is unique
-        expr = Expr(:block, copy(expr))
-        push!(expr.args, :(__pkguuid__ = $idmod))
-        # Add the expression to the fileinfo
-        exsnew = ExprsSigs()
-        exsnew[expr] = nothing
-        mexsnew = ModuleExprsSigs(modcaller=>exsnew)
-        # Before executing the expression we need to set the load path appropriately
-        prev = Base.source_path(nothing)
-        tls = task_local_storage()
-        tls[:SOURCE_PATH] = sourcefile
-        # Now execute the expression
-        mexsnew, includes = try
-            eval_new!(mexsnew, fi.modexsigs)
-        finally
-            if prev === nothing
-                delete!(tls, :SOURCE_PATH)
-            else
-                tls[:SOURCE_PATH] = prev
+        lock(pkgdatas_lock)   # ensure watch_package has finished
+        try
+            # Get/create the FileInfo specifically for tracking @require blocks
+            pkgdata = pkgdatas[id]
+            filekey = relpath(sourcefile, pkgdata) * "__@require__"
+            fileidx = fileindex(pkgdata, filekey)
+            if fileidx === nothing
+                files = srcfiles(pkgdata)
+                fileidx = length(files) + 1
+                push!(files, filekey)
+                push!(pkgdata.fileinfos, FileInfo(modcaller))
+            end
+            fi = pkgdata.fileinfos[fileidx]
+            # Tag the expr to ensure it is unique
+            expr = Expr(:block, copy(expr))
+            push!(expr.args, :(__pkguuid__ = $idmod))
+            # Add the expression to the fileinfo
+            exsnew = ExprsSigs()
+            exsnew[expr] = nothing
+            mexsnew = ModuleExprsSigs(modcaller=>exsnew)
+            # Before executing the expression we need to set the load path appropriately
+            prev = Base.source_path(nothing)
+            tls = task_local_storage()
+            tls[:SOURCE_PATH] = sourcefile
+            # Now execute the expression
+            mexsnew, includes = try
+                eval_new!(mexsnew, fi.modexsigs)
```

Comment:
The lock is fairly coarse. What happens here if this includes some top-level code (unsure if that can happen) and by some unfortunate series of events you need to aquire the lock again (but maybe on a different thread?)  That would have the potential for a deadlock.

Reply:
You're right that it encompasses more code than necessary. Thanks. I think there might be another way to do it, let me play with an alternative.

Reply:
Might be better in #491? Would appreciate your insights if you have time.